odoo/models.py

```diff
@@ -5590,35 +5590,68 @@ def exists(self):
         valid_ids = set([r[0] for r in self._cr.fetchall()] + new_ids)
         return self.browse(i for i in self._ids if i in valid_ids)
 
-    def _check_recursion(self, parent=None):
+    def _check_recursion(self, field_name=None):
         """
-        Verifies that there is no loop in a hierarchical structure of records,
-        by following the parent relationship using the **parent** field until a
-        loop is detected or until a top-level record is found.
+        Verifies that these records are not in any loop
+        by following the given relationship of the field.
+        By default the **parent** field is used as the relationship.
 
-        :param parent: optional parent field name (default: ``self._parent_name``)
+        Since EXCLUSIVE LOCK is not added for sake of performance,
+        loops may be created by concurrent transactions.
+
+        :param field_name: optional field name (default: ``self._parent_name``)
         :return: **True** if no loop was found, **False** otherwise.
         """
-        if not parent:
-            parent = self._parent_name
+        if not field_name:
+            field_name = self._parent_name
+        field = self._fields.get(field_name)
+        if not field:
+            raise ValueError('invalid field_name: %r' % (field_name,))
+        if not (
+            field.type in ('many2many', 'many2one')
+            and field.comodel_name == self._name
+            and field.store
+        ):
+            raise ValueError('field must be a many2one or many2many relation on itself: %r' % (field_name,))
+        if not self.ids:
+            return True
 
-        # must ignore 'active' flag, ir.rules, etc. => direct SQL query
+        # must ignore 'active' flag, ir.rules, etc.
+        # direct recursive SQL query with cycle detection for performance
+        self.flush_model([field_name])
+        if field.type == 'many2many':
+            relation = field.relation
+            col1 = field.column1
+            col2 = field.column2
+        else:
+            relation = self._table
+            col1 = 'id'
+            col2 = field_name
         cr = self._cr
-        self.flush_model([parent])
-        for id in self.ids:
-            current_id = id
-            seen_ids = {current_id}
-            while current_id:
-                cr.execute(SQL(
-                    "SELECT %s FROM %s WHERE id = %s",
-                    SQL.identifier(parent), SQL.identifier(self._table), current_id,
-                ))
-                result = cr.fetchone()
-                current_id = result[0] if result else None
-                if current_id in seen_ids:
-                    return False
-                seen_ids.add(current_id)
-        return True
+        cr.execute(SQL(
+            """
+            WITH RECURSIVE __reachability AS (
+                SELECT %(col1)s AS start, %(col2)s AS reach
```

Comment:
The term "reach" feels inappropriate in this context (especially as a noun).  I have never seen it used like that.  For the columns, I would use common pairs of nouns like start/stop, or source/destination.